### DISCUSSION TITLE
Allow User to Prefer Decoupled Clock Timing

### DISCUSSION THREAD

**Discussion by yoyyoy** [ppy/osu#31128] [Ideas] [answered]
a couple months ago, I started experiencing a bug where the hit objects on screen would be desynced from the game's music. I instantly recognized this as the same issue I was experiencing before August of 2023. That patch fixed logic with the game's clocks that would desync the notes. 

what made the issue resurface was me switching from my monitor's audio output, to my motherboard's audio output. the former used nvidia's audio drivers, and the latter used windows default "High Definition Audio Device". To date, I have been unable to recreate the issue on the nvidia drivers. But, the problem with going back to the nvidia drivers is that they don't get a benefit from using REAL.exe. And I would like to minimize my audio latency if possible. switching audio outputs in game would also fix it, as it would reinitialize BASS, but quickly gets out of sync again.

While this is almost certainly a driver or BASS issue, in my opinion we should try to mitigate it by using the decoupled clock already present in the game if desired by the player. I have three major reasons for why we should take this approach. 

First, this is not an isolated incident with just me having a problem. Here is a compilation of other players experiencing the same issue.
https://github.com/ppy/osu/issues/26237
https://github.com/ppy/osu/issues/28595
https://youtu.be/wwslcIOy0QI?si=zKk__KXO-8Ol0SuH&t=776

plus several more that i've lost the links to. it's seems like there's a common misconception that this is an audio latency issue, as it sounds like the hitsounds are off. but in reality it is the circles that are off.

Second, trying to solve this driver or BASS issue would be exceedingly difficult. I have tried to fix this for myself and have had no such luck. I've found some instances of this or something similar happening to other people on the internet, but no concrete solution. 

third, this seems to be how osu! stable already does its clocks. in this video, I compare osu!stable cutting edge, the current osu!lazer version, and my local development version with a change to always use the decoupled clock.
https://youtu.be/G7EZaivVksE

I worked on the local dev version a bit more and ended up with this, showing a real time example of how the clocks and vary quite a bit.
https://youtu.be/Kr-cBAtzmMI

the code for this latest version looks like this, + some other stuff to add the config option nothing crazy
in osu.Framework.Timing.DecoupingFramedClock.cs in ProcessFrame()
```
// We can only begin tracking time from the second frame, as we need an elapsed real time reference.
if (lastReferenceTime != null)
{
    double elapsedReferenceTime = (realtimeReferenceClock.CurrentTime - lastReferenceTime.Value) * Rate;

    decoupledCurrentTime += elapsedReferenceTime;
}

// If the source is running, use it unless a decoupled clock is preferred by the user
if (Source.IsRunning && (!PreferDecoupledClock?.Value ?? true))
{
    currentTime = Source.CurrentTime;
    shouldBeRunning = true;
    return;
}

// If we're not allowed to decouple, we should also just pass-through the source time.
if (!AllowDecoupling)
{
    currentTime = Source.CurrentTime;

    //osu! cookie depends on IsRunning being true to sync with beats
    shouldBeRunning = false || (PreferDecoupledClock?.Value ?? false);
    return;
}

// We then want to check whether our internal running state permits time to elapse in decoupled mode.
if (!shouldBeRunning)
    return;

currentTime = decoupledCurrentTime;
```

I first wanted to fix this at the source, by fixing whatever was causing the time to be reported incorrectly from the Source. but found that it calls a function called ChannelGetPosition in ppy.ManagedBass.Mix.BassMix. looks like calls to BASS and from the looks of it the BASS folks are stumped too. https://www.un4seen.com/forum/?topic=19806.0

**Comment by peppy** [comment]
What you propose will allow audio to drift away from the actual song for actual users. We don't want this to happen. I played many a rhythm game where this was a common thing. Because of the nature of non-realtime systems there will be a different offset each time you start playing, which will never be resolved. I have good memories of restarting gameplay many times until the offset aligned well enough to get through a level.

Sure, it might fix your broken case where your system is reporting drifting times in the first place, but it's not the fix I'm looking for.

I think we'd sooner switch audio libraries (something that is already working in a code state) than jump to ignoring the audio subsystem's reported times. Users which have the weird drift issue have reported it is fixed when using the alternative audio engine, so there's that.

**Reply by yoyyoy** [reply]
may I ask how stable's and lazer's implementations of beatmap clocks differ? I can't replicate the issue on stable so it's a lazer specific regression

**Reply by peppy** [reply, marked as answer]
we believe it may be a bug in bassmix, which we use in lazer but not stable. no one on the team can repro it yet unfortunately..
